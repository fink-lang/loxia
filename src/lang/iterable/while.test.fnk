{fink2js} = import '../../testing/generate.fnk'
{describe, it, expect, to_match_snapshot} = import '@fink/jest/test.fnk'


describe 'while', fn:
  it 'compiles', fn:
    expect
      fink2js '
        while item:
          item < 10
      '
      to_match_snapshot


  it 'compiles with accu', fn:
    expect
      fink2js '
        while _, cntr=0:
          (cntr < 2, cntr + 1)

        while , cntr=0:
          (cntr < 2, cntr + 1)

        while item, cntr=0:
          (item and cntr < 2, cntr + 1)
      '
      to_match_snapshot


  it 'compiles destructuring', fn:
    expect
      fink2js '
        while [foo, ...bar]:
          foo_alone = foo and is_empty bar
          foo_alone

        while [...foo, bar]:
          bar and is_empty foo
      '
      to_match_snapshot



describe 'while async', fn:
  it 'compiles', fn:
    expect
      fink2js '
        while item:
          await item < 10
      '
      to_match_snapshot
