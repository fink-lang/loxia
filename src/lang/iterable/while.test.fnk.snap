// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`while compiles 1`] = `
"(function* filter_while(ˆitems_2) {
  for (const ˆitem_1 of ˆitems_2) {
    const item = ˆitem_1;
    const ˆresult_3 = item < 10;
    if (ˆresult_3 !== true) return;
    yield ˆitem_1;
  }
});

Object.assign(module.exports, {});"
`;

exports[`while compiles destructuring 1`] = `
"(function* filter_while(ˆitems_2) {
  for (const ˆitem_1 of ˆitems_2) {
    const [foo, ...bar] = ˆitem_1;
    const ˆresult_3 = foo && bar.length === 0;
    if (ˆresult_3 !== true) return;
    yield ˆitem_1;
  }
});

(function* filter_while(ˆitems_6) {
  for (const ˆitem_4 of ˆitems_6) {
    let _do_result;

    {
      const [...ˆitems_5] = ˆitem_4;
      _do_result = [ˆitems_5.slice(0, -1), ˆitems_5.slice(-1)];
    }
    const [foo, [bar]] = _do_result;
    _do_result = undefined;
    const ˆresult_7 = bar && foo.length === 0;
    if (ˆresult_7 !== true) return;
    yield ˆitem_4;
  }
});

Object.assign(module.exports, {});"
`;
